design-proposals/vm-maintenance-mode.md

```diff
@@ -0,0 +1,57 @@
+# Overview
+Add a new "Maintenance" phase for the VMI/VM that represents a special state where the VM doesn't run and we can safely access the VM disks.
```

Comment:
This sounds a lot like the paused state (which currently isn't reported as a VMI phase, but rather as a condition).
Does this proposal aims to include any functional difference compared to the Paused state, or is it mostly about how this is reported via phases?

Reply:
@zcahana please correct me if I say something wrong. As far as I understand, the paused state is when the VM inside the pod is paused. In this case, it doesn't work for disk modification because you could modify a file from where the VM was reading/writing, and then when you unpause the data has changed. In our case, we don't want any running/paused instances of QEMU that could access the disk concurrently.

Reply:
I understand. In that case, how is that different than a Stopped VM?

One aspect that will need some attention is that for a Stopped VM, one or more PVCs might still be unprovisioned (in WaitForFirstConsumer state). That would normally happen before the VM is started for the first time, but I reckon hotplugged disks may introduce such WFFC PVCs at later lifecycle stages. But this on its own doesn't necessarily require the introduction of a new Maintenance phase.

Reply:
@zcahana the disk will be provisioned as far as we want to access them with the libguestfs pod because this pod needs to be started with the PVCs attached.

The Maintenace state differs from the Stopped VM because we refuse to start the VM if it is in this state. Additionally, we might want to place a Failed VM in Maintenance mode for example when there are some issues in booting and we want to examine the disks. This could be added to the stopped state, but it might be worthy to add a new phase to make the distinction. Right now I have also added the Maintenance state for VMIs but maybe this applies only to VMs.

I should also add in the proposal that in order to leave the maintenance phase we should check if there is a running libguestfs/maintenance pod.

Reply:
> the disk will be provisioned as far as we want to access them with the libguestfs pod because this pod needs to be started with the PVCs attached.

Right, but do notice that this might affect the VM scheduling afterwards, i.e., so we may want to scheduel the libguestfs pod with the same scheduling constraints as the VM, or somehow utilize the existing mechanism to create a temp pod for the VMI.

> The Maintenace state differs from the Stopped VM because we refuse to start the VM if it is in this state

That makes sense now. May be worthwhile to explain this in the proposal doc as well.

> Additionally, we might want to place a Failed VM in Maintenance mode for example when there are some issues in booting and we want to examine the disks

This has been partially introduced in kubevirt/kubevirt#6040, so implementation-wise it can make sense to connect these two capabilities.

Reply:
@zcahana thank! I need to add this and the scheduling affinity is an excellent suggestion!